Add JWT-SVID audience policy configuration with JTI claim support

srikalyan opened the pull request:
## Summary

Implements issue #6043 - JWT JTI Attribute support with per-audience policy configuration.

This PR introduces a flexible JWT-SVID policy system that allows operators to control:
- Whether JTI (JWT ID) claims are included in tokens
- Whether agent-side caching is used

### New `JWTSVIDAudiencePolicy` enum with three modes:
- **DEFAULT (0)**: No JTI claim, caching enabled. Backwards compatible behavior.
- **AUDITABLE (1)**: JTI claim included for audit trails, caching still enabled. Verifiers can optionally enforce replay protection.
- **UNIQUE (2)**: JTI claim included, caching disabled. Each request gets a fresh token for strict replay protection.

### Configuration options:
- `jwt_svid_default_audience_policy`: Default policy for audiences not explicitly configured
- `jwt_svid_audience_policies`: Per-audience policy overrides (map of audience → policy)

### Key behaviors:
- **Most restrictive wins**: When multiple audiences are requested, the strictest policy applies
- **Backwards compatible**: DEFAULT policy preserves existing behavior
- **Database schema v24**: Adds `entry_audience_policies` table for per-audience configuration

## What's included in this PR:
- Proto definitions for new enum and entry fields
- Server-side JTI generation based on effective policy
- Agent-side cache bypass for UNIQUE policy
- Database migration and CRUD operations
- Comprehensive tests

## What's NOT included (requires spire-api-sdk update):
- Entry API conversion logic
- CLI commands for configuring policies

These will follow in a separate PR after the spire-api-sdk is updated with the new types.

## Test plan
- [x] Unit tests for JTI generation with all policy modes
- [x] Unit tests for "most restrictive wins" multi-audience behavior
- [x] Unit tests for per-audience policy overrides
- [x] Database migration tests
- [x] Agent cache bypass tests for UNIQUE policy

Fixes #6043

srikalyan committed:
Add JWT-SVID audience policy configuration with JTI claim support (#6043)

Introduces per-audience JWT-SVID policy configuration to control JTI
(JWT ID) claim inclusion and agent-side caching behavior.

New JWTSVIDAudiencePolicy enum with three modes:
- DEFAULT (0): No JTI, caching enabled (backwards compatible)
- AUDITABLE (1): JTI included, caching enabled
- UNIQUE (2): JTI included, caching disabled

Features:
- Per-entry default policy via jwt_svid_default_audience_policy
- Per-audience overrides via jwt_svid_audience_policies map
- "Most restrictive wins" for multi-audience tokens
- Database schema v24 with entry_audience_policies table
- Agent cache bypass for UNIQUE policy

Note: API SDK and CLI integration will follow in a separate commit
after spire-api-sdk is updated.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

srikalyan committed:
Add CLI and entry conversion support for JWT-SVID audience policies

This commit adds support for the new JWT-SVID audience policy configuration
in the SPIRE server CLI and entry conversion logic:

CLI changes:
- Add -jwtSVIDDefaultAudiencePolicy flag for default audience policy
- Add -jwtSVIDAudiencePolicy flag for per-audience policy configuration
- Update entry create, update, and show commands to handle new fields
- Add AudiencePolicyFlag custom type for parsing audience:policy pairs

Entry conversion:
- Add JwtSvidDefaultAudiencePolicy and JwtSvidAudiencePolicies to
  EntryToProto and ProtoToEntry conversion functions
- Add audiencePolicyToInternal helper for enum conversion

Policy options: default, auditable, unique
- default: No JTI claim, caching enabled (current behavior)
- auditable: JTI claim included, caching enabled
- unique: JTI claim included, caching disabled (unique tokens)

Part of #6043

NOTE: Only merge after these dependent PRs are merged:
- spire-api-sdk: https://github.com/spiffe/spire-api-sdk/pull/84
- spire: https://github.com/spiffe/spire/pull/6514

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

srikalyan commented:
> Thanks @srikalyan for opening this! We discussed this today and we think it might be good to start with something simple at first by having just the entry level policy (maybe called something like `jwt_svid_jti_policy`). We think that should cover most uses. If it turns out there's still a need for the per audience version, we can add something like that later on.
> 
> We also think we should combine this flag with the changes in #6360. This avoid the need to have these seldomly used flags as individual columns in the database. I don't know what's the current state of the work from @ValFadeev, but maybe you can continue from where he left off?

@ValFadeev let me know if you are ok in me making the necessary change?

srikalyan commented:
@sorindumitru now that the other feature is merged, can we get this PR rolling?

sorindumitru commented:
Yes, I hope I can have a run over this over the weekend and provide some feedback.

sorindumitru commented:
> Thanks @srikalyan for opening this! We discussed this today and we think it might be good to start with something simple at first by having just the entry level policy (maybe called something like `jwt_svid_jti_policy`). We think that should cover most uses. If it turns out there's still a need for the per audience version, we can add something like that later on.

Had a look through the code for a bit, I'll look some more, but I notice we still have both the flag + the audience map. I don't think it's that likely to configure things per audience so in the interest of keeping the configuration simple we should start with a simple flag, maybe just `jws_svid_include_jti`?

You'll also have to add the flag to the new [AdditionalAttributes](https://github.com/spiffe/spire-api-sdk/blob/99d433165a01add2a207122dab97d3d1fa1f77ad/proto/spire/api/types/entry.proto#L79) structure instead of directly to the entry (this will also fix some issues around parsing the field in this PR, i think listing entries doesn't currently parse the fields).

srikalyan committed:
Bump spire-api-sdk to merged JTI commit (PR #95)

Replaces local-path replace directive with the merged upstream SDK
pseudo-version so CI can resolve the new jwt_svid_include_jti field
from the public module proxy.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

sorindumitru commented:
Hi @srikalyan, just checking in to see if you're still planning on updating this to use the new field added in the SDK.

srikalyan committed:
Persist AdditionalAttributes through entry update mask

The updateEntry mask translation in the entry v1 service was missing the
AdditionalAttributes field, so masked updates targeting that field were
silently dropped before reaching the datastore. Map the input mask bit
through and add coverage for both the honored and ignored cases.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

srikalyan committed:
Preserve sibling AdditionalAttributes on partial CLI entry update

The protobuf AdditionalAttributes message has no per-field mask, so a
CLI update that set only one of -disableX509SVIDPrefetch or
-jwtSVIDIncludeJTI would clobber the unset sibling bit with its zero
value. Track each flag's set state independently and, when exactly one
is supplied (and -data is not in use), fetch the existing entry to
merge the unspecified bit before BatchUpdateEntry. GetEntry failures
short-circuit before any update is attempted.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

srikalyan commented:
@sorindumitru sorry about it. I've made most changes and forgot to push them :). It should be up and ready for review now.

srikalyan commented:
@sorindumitru need your help on this

sorindumitru commented:
> @sorindumitru need your help on this

I'll try to get to this in the next couple of days, I just didn't have enough time to look at some things a bit more carefully.

sorindumitru commented on pkg/agent/manager/manager.go:
```diff
@@ -300,10 +300,18 @@ func (m *manager) FetchJWTSVID(ctx context.Context, entry *common.RegistrationEn
-	cachedSVID, ok := m.cache.GetJWTSVID(spiffeID, audience)
-	if ok && !m.c.RotationStrategy.JWTSVIDExpiresSoon(cachedSVID, now) {
-		return cachedSVID, nil
+	var cachedSVID *client.JWTSVID
+	var ok bool
+	if !bypassCache {
+		cachedSVID, ok = m.cache.GetJWTSVID(spiffeID, audience)
+		if ok && !m.c.RotationStrategy.JWTSVIDExpiresSoon(cachedSVID, now) {
+			return cachedSVID, nil
+		}
 	}
```
Would be good to have a test verify that the cache is not updated in the case we bypass the cache.

srikalyan replied:
Addressed it.

sorindumitru commented on pkg/server/api/entry/v1/service.go:
```diff
@@ -651,18 +651,19 @@ func (s *Service) updateEntry(ctx context.Context, e *types.Entry, inputMask *ty
+			X509SvidTtl:          inputMask.X509SvidTtl,
+			JwtSvidTtl:           inputMask.JwtSvidTtl,
+			Hint:                 inputMask.Hint,
+			AdditionalAttributes: inputMask.AdditionalAttributes,
```
Thanks for fixing this!

sorindumitru commented on pkg/server/api/svid/v1/service.go:
```diff
@@ -402,7 +403,7 @@ func (s *Service) NewJWTSVID(ctx context.Context, req *svidv1.NewJWTSVIDRequest)
 	}
 
-	jwtsvid, err := s.mintJWTSVID(ctx, entry.GetSpiffeId(), req.Audience, entry.GetJwtSvidTtl())
+	jwtsvid, err := s.mintJWTSVID(ctx, entry.GetSpiffeId(), req.Audience, entry.GetJwtSvidTtl(), entry.GetAdditionalAttributes().GetJwtSvidIncludeJti())
```
I think `GetAdditionalAttributes()` may return `nil`, so we probably have to check if it's nil or not. I don't know if we guarantee in another place that it ends being non-nil, but I think it's best to make sure it's not nil here.

Would  be good to have tests for this here, nil and non-nil additional attributes and the values of the JTI field.

srikalyan replied:
Fixed it.

sorindumitru commented on pkg/server/api/entry_test.go:
```diff
@@ -750,7 +750,8 @@ func TestReadOnlyEntryClone(t *testing.T) {
 		StoreSvid:      true,
 		AdditionalAttributes: &types.Entry_AdditionalAttributes{
-			DisableX509SvidPrefetch: false,
+			DisableX509SvidPrefetch: true,
+			JwtSvidIncludeJti:       true,
```
Did something change here that made us have to change the value for the prefetch bool?

srikalyan replied:
Sorry for the unintentional changes, reverted in the latest push. The only addition is JwtSvidIncludeJti: true; DisableX509SvidPrefetch is back to its prior value (false).

sorindumitru commented on pkg/server/credtemplate/builder.go:
```diff
@@ -347,6 +351,10 @@ func (b *Builder) BuildWorkloadJWTSVIDClaims(ctx context.Context, params Workloa
 	}
+
+	if params.IncludeJTI {
+		attributes.Claims["jti"] = uuid.Must(uuid.NewV4()).String()
```
I think this ends up with the `xxxx-xxx..-xxx` representation which is somewhat longer, mostly for human benefits. Maybe we can go with a `base64url` of the binary, which I think is smaller. If that's not smaller, what you have here also works.

sorindumitru commented on cmd/spire-server/cli/entry/update.go:
```diff
@@ -107,10 +110,16 @@ func (c *updateCommand) AppendFlags(f *flag.FlagSet) {
 	f.BoolFunc("disableX509SVIDPrefetch", "A boolean value that, when set, disables prefetching X509 SVID for this entry",
 		func(_ string) error {
-			c.additionalAttributesSet = true
+			c.disableX509SVIDPrefetchSet = true
 			c.disableX509SVIDPrefetch = true
 			return nil
 		})
```
I think it might be easier to reason about this if we also have `additionalAttributesSet` and always fetch the entry when it is set. Otherwise the more fields we introduce the more checks we have to do.

Maybe that's a sign that we should have a mask for it, but maybe that's for the future.

srikalyan replied:
Done. additionalAttributesSet now gates the fetch in Run(); per-flag ...set bits are kept only for the merge step (no per-field mask within AdditionalAttributes today). A future sub-mask would let us drop them.

sorindumitru reviewed (commented):
Sorry for the delay here, @srikalyan. I'll try to prioritise looking at the updates here.

srikalyan committed:
Address review feedback for JWT-SVID JTI claim

- pkg/server/api/svid/v1/service.go: add explicit nil check for
  GetAdditionalAttributes() before reading JwtSvidIncludeJti; add
  test cases covering nil attrs, JTI false, and JTI true paths.
- pkg/server/credtemplate/builder.go: return error from jti UUID
  generation instead of panicking via uuid.Must; encode the 16
  random bytes as base64url for a more compact claim value.
- cmd/spire-server/cli/entry/update.go: gate the existing-entry
  fetch on a single additionalAttributesSet flag, then merge each
  AdditionalAttributes field individually so unspecified flags
  preserve the existing value.
- pkg/agent/manager/manager_test.go: add TestFetchJWTSVIDWithJTIBypassesCache
  proving JwtSvidIncludeJti bypasses both the cache read and write.
- pkg/server/api/entry_test.go: only add JwtSvidIncludeJti: true
  in TestReadOnlyEntryClone; revert unintended prefetch change.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

sorindumitru reviewed (dismissed):
LGTM, thanks @srikalyan! Would you mind fixing up the DCO?

srikalyan committed:
Clone existing AdditionalAttributes on partial CLI entry update

Switch mergeAdditionalAttributes to proto.Clone the existing message and
override only fields explicitly set on the command line. This preserves
unknown fields and any future AdditionalAttributes the binary does not
know about, so the merge does not need editing when new fields are added.

Also harden the agent's bypassCache read with a nil-safe chained getter.

Signed-off-by: Srikalyan Swayampakula <srikalyansswayam@gmail.com>

srikalyan commented on cmd/spire-server/cli/entry/util_posix_test.go:
```diff
@@ -24,7 +24,8 @@ const (
     	Instance name to substitute into socket templates (env SPIRE_SERVER_PRIVATE_SOCKET_TEMPLATE).
-  -jwtSVIDTTL int
+  -jwtSVIDIncludeJTI
+` + "    \tA boolean value that, when set, includes a unique 'jti' claim in JWT-SVIDs issued for this entry and bypasses the agent JWT-SVID cache\n" + `  -jwtSVIDTTL int
```
Won't fix, cosmetic.

srikalyan commented:
@sorindumitru can you give one more ship it, had to address copilot review

sorindumitru reviewed (approved):
re-LGTM
